[time] [coqc] Print time for document setup and end.

ejgallego opened the pull request:
This can be useful when benchmarking.

As header in the time output, we use a special marker with a Coq
comment so it is still valid .v syntax: "(* Document Start / End *)"

SkySkimmer commented:
Timing the end of the document could also be nice, do you think you could include it in this PR?

ejgallego commented:
> Timing the end of the document could also be nice, do you think you could include it in this PR?

Yes, that's a good idea. Do you mainly mean `save_library_to` + (Stm.join) ?

SkySkimmer commented on toplevel/ccompile.ml:
```diff
@@ -20,6 +20,15 @@ let create_empty_file filename =
+  if opts.config.time then
+    let vernac = Vernacexpr.({ expr = VernacComments [CommentString header]; control = []; attrs = []}) in
+    let vernac = CAst.make ~loc vernac in
+    let header = Topfmt.pr_cmd_header vernac in
```
Shouldn't we expose some `pr_gen_header` which doesn't need to go through a vernacexpr instead?

ejgallego replied:
We could, but I deemed it not worth the effort.

SkySkimmer replied:
It bugs me so I pushed a commit

ejgallego replied:
Thanks!

SkySkimmer commented:
Test suite broken

ejgallego commented:
> Test suite broken

Fixed, note that I've updated the PR as not to produces a spurious "while in the prelude" line which is IMHO useless, moreover such stuff can be debugged just using `-noinit` + `From Coq Require Import Prelude`

gares commented:
shall we try a bench to see how the html turns out?

ejgallego commented:
Yes @gares , good idea.

@coqbot bench now

@SkySkimmer , the osx test suite stuff is fishy.

SkySkimmer commented:
For https://coq.gitlab.io/-/coq/-/jobs/3667396595/artifacts/_bench/html/coq-fiat-core/src/ADT.v.html we have
- OLD timing
~~~
Chars 0 - 45 [Require~Export~Fiat.ADT.ADTSig...] 0.168 secs (0.141u,0.027s)
~~~
- NEW timing
~~~
Chars 0 - 0 [Document~Start] 0.025 secs (0.019u,0.006s)
Chars 0 - 45 [Require~Export~Fiat.ADT.ADTSig...] 0.176 secs (0.152u,0.023s)
Chars 46 - 47 [Document~End] 0.034 secs (0.03u,0.003s)
~~~
- HTML
~~~html
<div class="code" title="File: ADT.v
Line: 1

Time1: 0.168s
Time2: 0.025s
">
<div class="time1" style="width: 95.454545454545%"></div>
<div class="time2" style="width: 14.204545454545%"></div>
<pre>Require Export Fiat.ADT.ADTSig Fiat.ADT.Core.</pre>
</div>
<div class="code" title="File: ADT.v
Line: 2

Time1: 0s
Time2: 0.176s
">
<div class="time1" style="width: 0%"></div>
<div class="time2" style="width: 100%"></div>
<pre>

</pre>
</div>
~~~

It got confused by the non matching outputs, we should run the bench with the new printing on both commits.

SkySkimmer commented:
@coqbot bench

gares commented:
This is good for testing future PR against master. But it will make it impossible to compare, say, Coq 8.17 with 8.18.
Was the error message understandable? If not it should be fixed.

gares commented:
Well, if it is not too hard to implement I'd prefer a message saying the two commits are not comparable because of this pr.

ejgallego commented:
Wouldn't it be possible for the function doing the diff to detect this case and have it working by just removing the 2 extra additions if the other diff doesn't have them?

ejgallego commented:
I don't have more time for this, if someone is interested in taking over I offer to do a rebase, otherwise I'll close.

JasonGross commented:
What was left to be done?

ejgallego commented:
> What was left to be done?

@JasonGross I think the last bit to do here was to fix the timing comparison functions so things work or bail out without producing trash when we have a timing made with a version that outputs the end / start timing data.

In a sense, this is related to the problem of comparing two timing datasets of the same document but a different versions. I'm quite interested in this for `coq-lsp`, as I'd like to add to the performance browser a setting to display performance differences among different `git` revisions of a given .v file.
